hooks/dhclient.in

```diff
@@ -0,0 +1,91 @@
+#! /bin/bash
+
+set -eu
+
+LXC_DHCP_SCRIPT="@LXCHOOKDIR@/dhclient-script"
+LXC_DHCP_CONFIG="@SYSCONFDIR@/lxc/dhclient.conf"
+
+rootfs_path="${LXC_ROOTFS_PATH#*:}"
+hookdir="${rootfs_path/%rootfs/hook}"
+
+conffile_arg=""
+if [ -e "${LXC_DHCP_CONFIG}" ]; then
+    conffile_arg="-cf ${LXC_DHCP_CONFIG}"
+fi
+
+debugfile="/dev/null"
+if [ "${LXC_LOG_LEVEL}" = "DEBUG" ] || [ "${LXC_LOG_LEVEL}" = "TRACE" ]; then
+    debugfile="${hookdir}/dhclient.log"
+    echo "INFO: Writing dhclient log at ${debugfile}." >&2
+fi
+
+pidfile="${hookdir}/dhclient.pid"
+leasefile="${hookdir}/dhclient.leases"
+
+usage() {
+    echo "Usage: ${0##*/} <name> lxc {start-host|stop}"
+}
+
+dhclient_start() {
+    mkdir -p "${hookdir}"
+
+    if [ -e "${pidfile}" ]; then
+        echo "WARN: DHCP client is already running, skipping start hook." >> "${debugfile}"
+    else
+        echo "INFO: Starting DHCP client and acquiring a lease..." >> "${debugfile}"
+        nsenter --uts --user --net --target "${LXC_PID}" -- \
```

Comment:
So I think this won't work unconditionally. The problem is that with `--user` the host and container need to be in different user namespaces. If they are not this will fail with `EINVAL`. I think what you want to do is look whether `LXC_USER_NS` is set if `LXC_HOOK_VERSION` is set to `1` or if `user:/proc/<pid>/fd/<nr>` is passed as argument when `LXC_HOOK_VERSION` is set to `0` or not set at all.

Reply:
Right, good catch. Shouldn't we do that for every namespace though since the user could have specified a `--share-<ns>`?

Reply:
If I remember correctly, then only user namespaces have this restriction. The other ones should be fine. Was that your concern or did I misunderstand?
(Fwiw, since I introduced `lxc.namespace.[namespace identifier]` users can actually even specify this in their config.)

Reply:
It works for the `stop` hook but what about the `start-host` hook? Do I have to check the userns proc inodes or there is a better way to know whether it's shared or not?

Reply:
One way to do it would be to compare:

    readlink /proc/${LXD_PID}/ns/user == readlink /proc/self/ns/user

Reply:
Since by the time the `start-hook` runs we already know the init pid of the container and we make it available in `LXC_PID`.

Reply:
Oh, and the `LXC_PID` is guaranteed to be valid until after the `start-host` hook is run since we place a barrier right after it. Only after the `start-hook`s are run is the the child allowed to `exec*()`.